### ISSUE TITLE
[Feature] How to support do_sample config just like Automodel 能否像Automodel推理中的do_sample参数支持，支持使用确定性生成方法，而不是随机采样

### ISSUE THREAD

**Issue by Leo-yang-1020** [InternLM/lmdeploy#1833]
### Motivation

When using lmdeploy to inference, sometimes we'd like to set do_sample = false, but according to official document there's no do_sampling config, can we add this just like Automodel? e.g: generation_config = dict(
    num_beams=1,
    max_new_tokens=512,
    do_sample=False,
)

**Comment by lvhan028** [comment]
We'll support `do_sample` in July. Currently, the team is rushing for the version in June.

**Comment by lvhan028** [comment]
`top_k=1` is equivalent to `do_sample=False`

**Comment by Leo-yang-1020** [comment]
> We'll support `do_sample` in July. Currently, the team is rushing for the version in June.

Thanks a lot!

**Comment by zhyncs** [comment]
Should our API design align with [Transformers](https://huggingface.co/docs/transformers/main_classes/text_generation) or OpenAI API?

ref
https://platform.openai.com/docs/api-reference/chat/create#chat-create-top_p
https://platform.openai.com/docs/api-reference/chat/create#chat-create-temperature

The OpenAI API does not have the `top_k` and `do_sample` parameters.

**Comment by lvhan028** [comment]
For the `pipeline` API, we would like to align with transformers

**Comment by zhjunqin** [comment]
> We'll support `do_sample` in July. Currently, the team is rushing for the version in June.

Does lmdeploy support do_sample now?

**Comment by lvhan028** [comment]
It is wrapped in PR https://github.com/InternLM/lmdeploy/pull/1966, which hasn't been reviewed yet.
We'll seperate that PR into smaller ones, and try to merge the `dp_sample` as soon as possible

**Comment by zhjunqin** [comment]
> It is wrapped in PR #1966, which hasn't been reviewed yet. We'll seperate that PR into smaller ones, and try to merge the `dp_sample` as soon as possible

Is there any build whl package that I can do test? Thanks.

**Comment by lvhan028** [comment]
Sorry, the answer is no. 

But you can set `top_k=1` or `temperature=0`  instead, which represents the greedy search method in lmdeploy

**Comment by meteorlium** [comment]
I need the model to output without sample and the prob of the first output token. Can it be implemented? If top_k=1 is used, prob cannot be output.

**Comment by KenForever1** [comment]
> Sorry, the answer is no.
> 
> But you can set `top_k=1` or `temperature=0` instead, which represents the greedy search method in lmdeploy

Hello, I want to set top_k=1 through the parameter of the [chat_completions_v1 function](https://github.com/InternLM/lmdeploy/blob/main/lmdeploy/serve/openai/api_client.py#L91), but I found that this parameter is not available. May I ask why this parameter is not added?
So I can set it up simultaneously, top_k=1、temperature=1.0， To achieve the effect of do_Sample=False. Because the server defaults to do_Sample=True.
```
if not gen_config.do_sample:
            logger.warning(f'GenerationConfig: {gen_config}')
            logger.warning(
                'Since v0.6.0, lmdeploy add `do_sample` in '
                'GenerationConfig. It defaults to False, meaning greedy '
                'decoding. Please set `do_sample=True` if sampling '
                ' decoding is needed')
            # greedy decode
            gen_config.top_k = 1
            # avoid unnecessary process
            gen_config.temperature = 1.0
            gen_config.repetition_penalty = 1.0
```

**Comment by KenForever1** [comment]
> May try `temperature=0.0`

Actually, I want to use Greedy sampling instead of Random_Sample by setting top_k=1. However, setting temperature to 0.0 cannot achieve this goal.

**Comment by lvhan028** [comment]
You can set `topk` in the extra body of payload.
@AllentDan Could you provide an example?

**Comment by AllentDan** [comment]
```
completion = client.chat.completions.create(
  model="NousResearch/Meta-Llama-3-8B-Instruct",
  messages=[
    {"role": "user", "content": "hi"}
  ],
  extra_body={
    "topk": 3
  }
)
```

**Comment by KenForever1** [comment]
> You can set `topk` in the extra body of payload. [@AllentDan](https://github.com/AllentDan) Could you provide an example?

Thanks for your reply. Using the extra_body parameter of OpenAI client may be useful. 
I am currently using lmdeploy.serve.openai.api_client APIClient, chat_completions_v1 function. I have manually added a top_k parameter, just curious why this function did not add the top_k parameter. Other APIs, such as the completions_v1 and chat_interactive_v1 function, have a top_k parameter. Perhaps you have design considerations.

**Comment by KenForever1** [comment]
> [@KenForever1](https://github.com/KenForever1) There is `top_k` there
> 
> [lmdeploy/lmdeploy/serve/openai/api_server.py](https://github.com/InternLM/lmdeploy/blob/3f8b079224d109aaa1b512867203a47ea600aa7d/lmdeploy/serve/openai/api_server.py#L393)
> 
> Line 393 in [3f8b079](/InternLM/lmdeploy/commit/3f8b079224d109aaa1b512867203a47ea600aa7d)
> 
>  top_k=request.top_k,

https://github.com/InternLM/lmdeploy/blob/3f8b079224d109aaa1b512867203a47ea600aa7d/lmdeploy/serve/openai/api_client.py#L97

Hello, The client parameter here does not have top_k, which may be a bug.

**Comment by AllentDan** [comment]
Yes, @KenForever1 it was missed in api_client. Thanks, we will fix it ASAP.
